Merge filter branch with main

Jelco-C opened the pull request:
Merge the filter branch with fully implemented filter_authors and filter_files functions with test suite into the main branch.

Jelco-C committed:
feat(analysis): Update filter_authors to use author_id

Update filter_authors to use author_id, reflecting the changes made to the 'Commit' shared type.

Jelco-C committed:
feat(filters): Create convert_authors_to_csv function

As the Author shared type contains nested fields, it must be flattened before conversion to CSV. This is taken care of in the convert_authors_to_csv function.

Jelco-C committed:
feat(filesystem): Create convert_authors_to_csv function

As the Author shared type contains nested fields, it must be flattened before conversion to CSV. This is taken care of in the convert_authors_to_csv function.

Jelco-C committed:
feat(analysis): Update test functions to match shared type definitions

Update the shared types initializations in test functions to match the new shared type definitions.

Jelco-C committed:
feat(analysis): Update test functions to match shared type definitions

Update the shared types initializations in test functions to match the new shared type definitions.

Jelco-C committed:
build(filesystem): Add Author import

Add Author shared type import for testing.

Jelco-C committed:
build(filters): Add Author import

Add Author shared type import for testing.

Jelco-C committed:
test(filters): Change repo test functions to match gitwrapper

Update the test_non_git_repo_should_fail() test function to match the new implementation of repo_discover using git_wrapper::Discover. As this function searches upward, modify the test function to verify this behavior.

Jelco-C committed:
test(filesystem): Change repo test functions to match gitwrapper

Update the test_non_git_repo_should_fail() test function to match the new implementation of repo_discover using git_wrapper::Discover. As this function searches upward, modify the test function to verify this behavior.

Jelco-C committed:
build(shared_types): Add traits to File, Line and LineType structs

Add partialeq, eq and hash traits to the File, Line and LineType structs as they are needed for the filter_authors implementation.

Jelco-C committed:
build(shared_types): Add traits to File, Line and LineType structs

Add partialeq, eq and hash traits to the File, Line and LineType structs as they are needed for the filter_authors implementation.

Jelco-C committed:
Merge remote-tracking branch 'origin/feature/filter-functions' into feature/filter-functions

Jelco-C committed:
feat(filters): Remove filter_metrics() function

Remove the filter_metrics() function, as all metrics are already filtered inside the filter_authors() and filter_files() functions.

Jelco-C committed:
feat(files): First implementation of the filter_files function

A first implementation of the filter_files() function, which removes file entries, updates commit metrics and author metrics.

Jelco-C committed:
test(files): Implement filter_files test function

Implement first version of a filter_files test function that tests whether all coresponding fields are updated in the test_analysis_result struct.

Jelco-C committed:
test(files): Test whether empty commits are removed by filter_files.

Implement test function to verify that commits with no files remaining after filtering are removed.

Jelco-C committed:
build(shared_types): Remove traits from file, line and linetype

Remove the hash, partialeq and eq traits from the structs. As files are now filtered based on a hashmap of their file IDs instead of the full files, these traits are no longer required.

Jelco-C committed:
feat(filesystem): Add recalculate_repository_metrics() helper function

Add a helper function to recalculate the repository-level metrics based on the file, commit and author-level metrics.

Jelco-C committed:
build(filters): Change filter_authors() signature to return repository

Change the filter_authors() function to return a repository instead of an analysisresult, which would contain redundant information. Corresponding to this change, also update the test functions.

Jelco-C committed:
feat(filters): Add repository metric calculations to filter functions

Add the recalculate_repository_metrics() function to filter_authors and replace the code inside filter_files with the new helper function.

Jelco-C committed:
test(filters): Test repository metric filtering of filter_authors

Add a test function to test whether the repository-level metrics are updated correctly by filter_authors.

impaex commented on gi_core/src/analysis.rs:
```diff
@@ -680,40 +680,149 @@ pub(crate) fn analyse_repository(params: &AnalysisParameters) -> Result<Reposito
-
-fn filter_files(result: AnalysisResult) {
-    // Placeholder for future implementation
+fn recalculate_repository_metrics(repository: &mut Repository) {
```
Even though updating the metrics using a separate function is more concise for us developers, I think it hurts performance, as we basically loop through the full repository twice (or more), once for the actual filtering, and once (or more) for updating the metrics.
Maybe doing this within the same loop would result in better performance?

impaex commented on gi_core/src/analysis.rs:
```diff
@@ -680,40 +680,149 @@ pub(crate) fn analyse_repository(params: &AnalysisParameters) -> Result<Reposito
 
-fn filter_metrics(result: AnalysisResult) {
-    // Placeholder for future implementation
+fn filter_authors(result: AnalysisResult, authors_to_exclude: Vec<Author>) -> Result<Repository, String> {
```
We might want to change the filter_authors signature to be:

```rust
fn filter_authors(repo: Repository, filter: Filter) -> Result<Repository, String>
```

As we do not want filter processing to happen in the API. 
We now have separate filter processing in each analysis function (can be refactored later, moving filter processing into one function which the analysis and filter functions can call), which this function should do too for now. Look at the main analysis functions for (pretty much copy-pastable) snippets on filter processing.

impaex commented on gi_core/src/analysis.rs:
```diff
@@ -680,40 +680,149 @@ pub(crate) fn analyse_repository(params: &AnalysisParameters) -> Result<Reposito
+    Ok(repository)
+}
+
+fn filter_files(result: AnalysisResult, files_to_exclude: Vec<File>) -> Result<Repository, String> {
```
We might want to change the filter_files signature to be:

```rust
fn filter_files(repo: Repository, filter: Filter) -> Result<Repository, String>
```

As we do not want filter processing to happen in the API. 
We now have separate filter processing in each analysis function (can be refactored later, moving filter processing into one function which the analysis and filter functions can call), which this function should do too for now. Look at the main analysis functions for (pretty much copy-pastable) snippets on filter processing.

impaex commented on gi_core/src/filesystem.rs:
```diff
@@ -74,6 +74,68 @@ pub fn convert_to_csv<T: serde::Serialize>(object_to_convert: &[T]) -> Result<St
         .map_err(|error| format!("Error converting CSV bytes to string: {}", error))
 }
 
+pub fn convert_authors_to_csv(authors: &[Author]) -> Result<String, String> {
```
Wouldn't it be better to move this functionality as a struct function?
Not sure if it would be, but feels out of place in the filesystem module.

Jelco-C committed:
build(filters): Change author_to_csv function into method of Author

Change the author_to_csv function into a method of the Author struct for cleaner organization.

Jelco-C committed:
feat(filter): Add the AuthorCsv struct for csv conversion method

Add the AuthorCsv struct which describes a flattened Author struct to be used in the Author::to_csv method.

Jelco-C committed:
build(filters): Change filter_files() signature and repo-level metrics

Update the filter_files() function to accept the Repository and Filter types as inputs. Also incorporate repo-level metric calculations to avoid looping through all data twice, as was the case for the previous implementation.

Jelco-C committed:
test(filters): Update existing filter_file tests to match new signature

Update the existing filter_file() test functions to match the new signature that accepts repository and filter as inputs.

Jelco-C committed:
test(filter): Add test function to include a Filter

Add a way to test whether inclusion of a Filter works.

Jelco-C committed:
build(filter): Change filter_author() input to Repository and Filter

Change the filter_author() function signature to accept Repository and Filter as an input.

Jelco-C committed:
test(filter): Update filter_author tests to match changes to signature

Update the already existing filter_author() test functions to match the signature change to accept Repository and Filter as inputs.

Jelco-C committed:
test(filter): Add pattern and wildcard filter_author test functions

Add a test function to test the filtering of patterns for authors and add a test function to filter using the wildcard character.

impaex committed:
Squashed commit of the following:

commit ee8969b306c4e2c2ada2dc5c874b65db2a132834
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Wed Dec 3 11:39:33 2025 +0100

    fix(gui): align author-file gui type with core type

commit f365fe28cbf0241b8406feeef0e3905bdb72a92a
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Wed Dec 3 11:02:25 2025 +0100

    refactor(gui): contrast beteween sticky author/file column and others

commit 0ac6a41f0299d420ff4831706aa688445a26678d
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Wed Dec 3 10:32:18 2025 +0100

    feat(gui): add search bar for found repos in data import

commit 13ddfb469ac8127ad76f189d6bfd40f720f533a9
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Wed Dec 3 10:20:33 2025 +0100

    feat(gui): add file explorer dialog for setting root path in data import

commit fd0dcdd094698837e4b63325721e0a593f0b27d9
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Wed Dec 3 09:41:03 2025 +0100

    refactor(gui): remove bar chart animation (not smooth for large repos)

commit 8a3c4cd663b35029c48a4bb76bd4577c272acd34
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Wed Dec 3 09:36:42 2025 +0100

    refactor(gui): show neutral icon when filters inactive for data import

commit d192ba1b74ac3eae756817f4c80c50aeb2f4242f
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Wed Dec 3 09:11:14 2025 +0100

    fix(gui): prevent table row from sliding into 'all authors/files' cell

commit 5aebeba504aad6b9f28d58baed74b34a9a9e80fb
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Wed Dec 3 09:10:15 2025 +0100

    feat(gui): add loading spinner overlay when calling core api

commit 6ee425f09eb7951006b001a49ffe4747f28cef42
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Dec 2 21:46:59 2025 +0100

    refactor(gui): change in last modified metric format

commit 594bc40581c248a34bf0b3d0440bdb89fcb1f046
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Dec 2 21:03:44 2025 +0100

    refactor(gui): change 'value' in pie chart tooltip to value name

commit 66d9eb427e6739d0522b596ba457b3b467661a28
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Dec 2 20:50:47 2025 +0100

    refactor(gui): small style changes to data import alertdialog window

commit fccbfb5365386b52e552201366ecaa2c192f3ee5
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Dec 2 20:05:15 2025 +0100

    refactor(gui): align filter data toggle with other buttons in sidebar

commit 4df4cd55f0eb682ab331d46a286b7b37bdeb70a9
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Dec 2 20:04:50 2025 +0100

    feat(gui): add button to data import alertdialog to main window

commit a0e44148396cb4b6dcfb63effc10ef634cfa267a
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Dec 2 20:04:17 2025 +0100

    refactor(gui): unite data import and analysed repo sections of sidebar

commit 80f1cda621234bd0c15eadf2964f274093a4b1b4
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Dec 2 20:02:22 2025 +0100

    refactor(gui): create separate component for data import alertdialog

commit 30e9796dd0e5aac397f976d8c9a4e96b9d20f8c8
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Dec 2 19:58:47 2025 +0100

    refactor(gui): add scrollable field to commit filter when no repo set

commit 96dfe25ee2dc58c4e238fd93b6269ebb877263ca
Author: Max <max.den.oudsten@gmail.com>
Date:   Mon Dec 1 13:14:47 2025 +0100

    feat(analysis): gives filter functions new signature

    These need to be properly implemented still

commit a64e5f985e54ae3af79edaf10c8aa4b80d9e98ad
Author: Max <max.den.oudsten@gmail.com>
Date:   Mon Dec 1 13:14:17 2025 +0100

    feat(API): rewrites rerun_analysis to use specific filter functions to increase processing speed

commit 4425b7828679b4b1d62c3053e0b1d80634f65929
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Nov 25 13:32:22 2025 +0100

    feat(gui): load settings using core api after setting path

commit dde6be53046d50698bdfc43ef6b82472e07051d3
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Nov 25 09:58:10 2025 +0100

    eat(gui): implement load settings json as params  for initial analysis

commit 553edb03b6d3fd44929ba88e234bdf1ad52511b2
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Nov 25 09:19:56 2025 +0100

    refactor(gui): prevent bar x-axis overflow in author viz

commit 61cc64adb350ff02696f369b72a8f9c2765ab604
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Nov 25 09:19:23 2025 +0100

    feat(gui): implement file/commit range filtering through core api

commit 32aaa2864e8d189852a2c1350a1d20ccd9ae164f
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Nov 25 09:18:52 2025 +0100

    refactor(gui): group files per folder in files filter in sidebar

commit a9905b28e470fdc0acef4a1b44c25693251527ba
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Nov 25 09:18:21 2025 +0100

    feat(gui): add functionality for filter data/no filtering toggle

commit 542aa7ae66cff6505c98b45e7dcb18b509a93eaa
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Tue Nov 25 09:17:51 2025 +0100

    feat(gui): implement settings load/save through core api

commit 7cf465d8091621ce63f6ab2afab776bb3f87608b
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Mon Nov 24 15:59:45 2025 +0100

    fix(gui): metric aggregation per day in author viz

commit 4a05d6ac2c154aaff4d3fbdb495e81bec05c80e5
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Sun Nov 23 18:35:02 2025 +0100

    feature(gui): implemented author filtering by analysis rerun core call

commit 67d486e6888c979a32861ccba6e0ed5338cf567e
Author: Max <max.den.oudsten@gmail.com>
Date:   Sun Nov 23 15:23:29 2025 +0100

    test(analysis): adds testcases further enlighting and testing glob matcher behaviour

commit b7d5d5620215584f88d10ba99ed6761132f4d827
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Sun Nov 23 09:45:36 2025 +0100

    refactor(gui): remove console debug output

commit c64c4922d86a5b28f9232da826bcbe922d32175d
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Sun Nov 23 08:37:59 2025 +0100

    feat(gui): add option to select which of found repos to add to analysis

    User now receives feedback on whether repos were found in the selected path and can select which ones to include for further analysis. User can also combine repos from multiple paths. User can also delete the full list of repos.

commit 849714f41b02b302a86deeea1f76e6a58747a63f
Author: Arlette9050 <90265412+Arlette9050@users.noreply.github.com>
Date:   Sun Nov 23 03:18:15 2025 +0100

    refactor(gui): import data from core api (instead of json example files)

commit b7dfbcd8b403cfcb5ec2bd20b58e3528ee1b6a9c
Author: Max <max.den.oudsten@gmail.com>
Date:   Fri Nov 21 11:24:03 2025 +0100

    test(analysis): adds unit tests for the new filters

commit 8997247c3f8e4234cab30f1c9521516926fab844
Author: Max <max.den.oudsten@gmail.com>
Date:   Fri Nov 21 11:23:49 2025 +0100

    fix(API): adds missing filters to API functions

commit de2ea110ed86b740870883ee79ccd28f2fbf1a2d
Author: Max <max.den.oudsten@gmail.com>
Date:   Fri Nov 21 11:11:30 2025 +0100

    feat(shared_types): adds author email and author name filters

commit 257d035ef517af50c211b3de2589137c60fc29b2
Author: Max <max.den.oudsten@gmail.com>
Date:   Fri Nov 21 11:11:14 2025 +0100

    feat(analysis): adds author email and author name filters

commit 1e2a8ccf7fd4830dbd211318b55bd20edafd9b20
Author: Max <max.den.oudsten@gmail.com>
Date:   Thu Nov 20 12:44:34 2025 +0100

    feat(data_import): Adds testing / demo code

    Open the console of the app (right click on canvas > Inspect > Console) to see the output of the API functions

commit fa65a7952826d7497c14f568bca4c90de2147b9f
Author: Max <max.den.oudsten@gmail.com>
Date:   Thu Nov 20 12:43:45 2025 +0100

    feat: adds TS API for easy function calling in frontend

commit e25c15820222633529f1cf0490264771142bf1fb
Author: Max <max.den.oudsten@gmail.com>
Date:   Thu Nov 20 12:43:16 2025 +0100

    feat(tauri): registers remaining core API functions as tauri commands

commit b84d471a1d90c2c68dfc9d11ef64519ceaf73f2f
Author: Max <max.den.oudsten@gmail.com>
Date:   Thu Nov 20 10:42:09 2025 +0100

    feat(gui): registers core API as tauri commands

commit 34bf6f74f207c46b2ea8fad9f9fba58da02735e4
Author: Max <max.den.oudsten@gmail.com>
Date:   Thu Nov 20 10:41:54 2025 +0100

    build(gui): adds core as dependency

commit 2832bf759424c2bbdc41b2361eee1114552eda81
Author: Max <max.den.oudsten@gmail.com>
Date:   Thu Nov 20 10:26:15 2025 +0100

    feat(filesystem): implements much faster git repository lookup

impaex committed:
fix: addresses Max's signature mistake

This push corrects my fault in which I gave an improper function signature, resulting in an implementation that was off.

impaex committed:
test(filesystem): fixes failing test regarding upward repo discovery

Jelco-C committed:
Merge branch 'main' into feature/filter-functions
